typescript/promisify-anything/src/query-athena.ts

```diff
@@ -0,0 +1,56 @@
+import * as restate from "@restatedev/restate-sdk";
+import * as athena from "@aws-sdk/client-athena";
+
+const QUERY_SQL = `
+    SELECT sum("table"."value") AS answer
+    FROM "demo_db"."table"
+`;
+
+export default async function queryAthena(ctx: restate.RpcContext, query?: string) {
+  const requestId = ctx.rand.uuidv4(); // make up a stable idempotency token for calling Athena API
+
+  const client = new athena.AthenaClient({});
+  const startQueryOutput = await ctx.sideEffect(
+    terminateOnError(
+      (error) => (error instanceof athena.AthenaServiceException ? error.$fault === "client" : false),
+      async () =>
+        await client.send(
+          new athena.StartQueryExecutionCommand({
+            QueryString: query ?? QUERY_SQL,
+            WorkGroup: "demo-workgroup",
+            ClientRequestToken: requestId,
+          }),
+        ),
+    ),
+  );
+  const queryId = { QueryExecutionId: startQueryOutput.QueryExecutionId };
+
+  const state = await ctx.sideEffect(async () =>
+    assertFinalState((await client.send(new athena.GetQueryExecutionCommand(queryId))).QueryExecution?.Status?.State),
+  );
```

Comment:
@tillrohrmann what do you think about this - as an alternative to a for-loop or throwing inside the sideEffect body?

Reply:
Generally looks good. What might not be so easy to understand is that we are using Restate's retry functionality to wait for the query to complete. Maybe adding something like ". Retrying." to the message of the Error in `assertFinalState` can make this clearer?